File: java/docspec-processor-java/src/main/java/io/docspec/processor/dsti/NamingAnalyzer.java
package io.docspec.processor.dsti;

import io.docspec.annotation.DocDeterministic;
import io.docspec.annotation.DocMethod;

/**
 * Parses method names to extract semantic verbs and objects using camelCase conventions.
 * Used by the DSTI (Documentation Semantic & Temporal Intelligence) system to infer
 * developer intent from naming patterns.
 */
public class NamingAnalyzer {

    public record NameSemantics(String verb, String object, String intent) {}

    /** Known verb prefixes, ordered longest-first where needed to avoid prefix collisions. */
    private static final String[] VERBS = {
            "get", "set", "is", "has", "find", "create", "delete", "remove",
            "update", "save", "add", "validate", "check", "compute", "calculate",
            "process", "handle", "convert", "transform", "parse", "format",
            "build", "generate", "initialize", "init", "load", "fetch",
            "send", "receive", "publish", "subscribe", "notify", "dispatch",
            "sync", "migrate", "schedule", "retry", "batch",
            "aggregate", "merge", "split", "emit", "enrich", "filter"
    };

    /**
     * Analyzes a method name and extracts its semantic components.
     *
     * @param methodName the simple name of the method
     * @return a {@link NameSemantics} record with verb, object, and intent category
     */
    @DocDeterministic
    @DocMethod(since = "3.0.0")
    public NameSemantics analyze(String methodName) {
        // Extract verb prefix
        for (String verb : VERBS) {
            if (methodName.startsWith(verb) && methodName.length() > verb.length()) {
                char nextChar = methodName.charAt(verb.length());
                if (Character.isUpperCase(nextChar)) {
                    String object = methodName.substring(verb.length());
                    String intent = categorizeIntent(verb);
                    return new NameSemantics(verb, splitCamelCase(object), intent);
                }
            }
        }
        // Fallback: the entire name is the verb
        return new NameSemantics(methodName, "", "unknown");
    }

    private String categorizeIntent(String verb) {
        return switch (verb) {
            case "get", "find", "fetch", "load", "is", "has" -> "query";
            case "set", "update", "save", "add" -> "mutation";
            case "create", "build", "generate", "initialize", "init" -> "creation";
            case "delete", "remove" -> "deletion";
            case "validate", "check" -> "validation";
            case "compute", "calculate", "process" -> "computation";
            case "convert", "parse", "format" -> "transformation";
            case "handle", "dispatch" -> "handler";
            case "send" -> "emission";
            case "receive" -> "consumption";
            case "sync" -> "synchronize";
            case "migrate" -> "migrate";
            case "schedule" -> "schedule";
            case "retry" -> "retry";
            case "batch" -> "batch-process";
            case "aggregate" -> "aggregate";
            case "merge" -> "merge";
            case "split" -> "split";
            case "notify" -> "notify";
            case "emit" -> "emit";
            case "publish" -> "publish";
            case "subscribe" -> "subscribe";
            case "transform" -> "transform";
            case "enrich" -> "enrich";
            case "filter" -> "filter";
            default -> "unknown";
        };
    }

    private String splitCamelCase(String s) {
        // Insert spaces before uppercase letters
        return s.replaceAll("([a-z])([A-Z])", "$1 $2").toLowerCase().trim();
    }
}

File: csharp/processor/Dsti/NamingAnalyzer.cs
using System.Linq;
using System.Text.RegularExpressions;

namespace DocSpec.Analyzer.Dsti;

// [DocModule(Id = "docspec-csharp-naming-analyzer", Name = "DSTI Naming Analyzer")]
// [DocDescription("Parses method names to extract semantic verbs and objects using PascalCase conventions.
//   Supports 40+ verbs across 20+ intent categories. C# equivalent of Java's NamingAnalyzer.")]
// [DocSince("3.0.0")]

/// <summary>
/// Parses method names to extract semantic verbs and objects using PascalCase/camelCase conventions.
/// Used by the DSTI (Deep Structural and Textual Intent) system to infer
/// developer intent from naming patterns. The C# equivalent of Java's <c>NamingAnalyzer</c>.
/// </summary>
// [DocDeterministic]
// [DocIntentional("Parse PascalCase method names into verb-object-intent triples for DSTI Channel 1")]
public class NamingAnalyzer
{
    /// <summary>Result of a name analysis containing verb, object, and inferred intent category.</summary>
    public record NameSemantics(string Verb, string Object, string Intent);

    /// <summary>Known verb prefixes, ordered longest-first where needed to avoid prefix collisions.</summary>
    private static readonly string[] Verbs =
    {
        "Get", "Set", "Is", "Has", "Find", "Create", "Delete", "Remove",
        "Update", "Save", "Add", "Validate", "Check", "Compute", "Calculate",
        "Process", "Handle", "Convert", "Transform", "Parse", "Format",
        "Build", "Generate", "Initialize", "Init", "Load", "Fetch",
        "Send", "Receive", "Publish", "Subscribe", "Notify", "Dispatch",
        "Sync", "Migrate", "Schedule", "Retry", "Batch",
        "Aggregate", "Merge", "Split", "Emit", "Enrich", "Filter",
        // C#-specific patterns
        "Try", "Begin", "End", "On", "Invoke", "Execute", "Run",
        "Map", "Resolve", "Register", "Configure", "Setup"
    };

    /// <summary>
    /// Analyzes a method name and extracts its semantic components.
    /// </summary>
    /// <param name="methodName">The simple name of the method (PascalCase in C#).</param>
    /// <returns>A <see cref="NameSemantics"/> record with verb, object, and intent category.</returns>
    // [DocMethod(Since = "3.0.0")]
    // [DocDeterministic]
    // [DocPreserves("verb-object separation from PascalCase")]
    public NameSemantics Analyze(string methodName)
    {
        // Try each known verb prefix
        foreach (var verb in Verbs)
        {
            if (!methodName.StartsWith(verb) || methodName.Length <= verb.Length) continue;

            char nextChar = methodName[verb.Length];
            if (char.IsUpper(nextChar) || (verb.Length == methodName.Length))
            {
                string obj = methodName.Substring(verb.Length);
                string intent = CategorizeIntent(verb.ToLowerInvariant());
                return new NameSemantics(verb.ToLowerInvariant(), SplitCamelCase(obj), intent);
            }
        }

        // Fallback: the entire name is the verb
        return new NameSemantics(methodName.ToLowerInvariant(), "", "unknown");
    }

    // [DocDeterministic]
    private static string CategorizeIntent(string verb)
    {
        return verb switch
        {
            "get" or "find" or "fetch" or "load" or "is" or "has" => "query",
            "set" or "update" or "save" or "add" => "mutation",
            "create" or "build" or "generate" or "initialize" or "init" => "creation",
            "delete" or "remove" => "deletion",
            "validate" or "check" => "validation",
            "compute" or "calculate" or "process" => "computation",
            "convert" or "parse" or "format" => "transformation",
            "handle" or "dispatch" or "on" => "handler",
            "send" or "emit" or "publish" => "emission",
            "receive" or "subscribe" => "consumption",
            "sync" => "synchronize",
            "migrate" => "migrate",
            "schedule" => "schedule",
            "retry" => "retry",
            "batch" => "batch-process",
            "aggregate" => "aggregate",
            "merge" => "merge",
            "split" => "split",
            "notify" => "notify",
            "transform" => "transform",
            "enrich" => "enrich",
            "filter" => "filter",
            "try" => "attempt",
            "begin" or "end" => "lifecycle",
            "invoke" or "execute" or "run" => "execution",
            "map" => "mapping",
            "resolve" => "resolution",
            "register" or "configure" or "setup" => "configuration",
            _ => "unknown"
        };
    }

    // [DocDeterministic]
    private static string SplitCamelCase(string s)
    {
        // Insert spaces before uppercase letters: "UserAccount" -> "user account"
        return Regex.Replace(s, "([a-z])([A-Z])", "$1 $2").ToLowerInvariant().Trim();
    }
}
